java/ql/src/utils/model-generator/CaptureSummaryModels.ql

```diff
@@ -0,0 +1,233 @@
+ * @id java/utils/model-generator/summary-models
+ */
+
+import java
+import ModelGeneratorUtils
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.internal.DataFlowImplCommon
+import semmle.code.java.dataflow.internal.DataFlowNodes
+import ModelGeneratorUtils
+
+string captureFlow(Callable api) {
+  result = captureQualifierFlow(api) or
+  result = captureParameterFlowToReturnValue(api) or
+  result = captureFieldFlowIn(api) or
+  result = captureParameterToParameterFlow(api) or
+  result = captureFieldFlow(api)
+}
+
+/**
+ * Capture fluent APIs that return `this`.
+ * Example of a fluent API:
+ * ```
+ * public class Foo {
+ *   public Foo someAPI() {
+ *    // some side-effect
+ *    return this;
+ *  }
+ * }
+ * ```
+ */
+string captureQualifierFlow(Callable api) {
+  exists(ReturnStmt rtn |
+    rtn.getEnclosingCallable() = api and
+    rtn.getResult().(ThisAccess).isOwnInstanceAccess()
+  ) and
+  result = asValueModel(api, "Argument[-1]", "ReturnValue")
+}
+
+/**
+ * Capture APIs that return tainted instance data.
+ * Example of an API that returns tainted instance data:
+ * ```
+ * public class Foo {
+ *   private String tainted;
+ *
+ *   public String returnsTainted() {
+ *     return tainted;
+ *   }
+ *
+ *   public void putsTaintIntoParameter(List<String> foo) {
+ *     foo.add(tainted);
+ *   }
+ * }
+ * ```
+ * Captured Model:
+ * ```
+ * p;Foo;true;returnsTainted;;Argument[-1];ReturnValue;taint
+ * p;Foo;true;putsTaintIntoParameter;(List);Argument[-1];ReturnValue;taint
+ * ```
+ */
+string captureFieldFlow(Callable api) {
+  exists(FieldAccess fa, ReturnNodeExt returnNode |
+    not (fa.getField().isStatic() and fa.getField().isFinal()) and
+    returnNode.getEnclosingCallable() = api and
+    fa.getCompilationUnit() = api.getCompilationUnit() and
+    isRelevantType(api.getReturnType()) and
+    not api.getDeclaringType() instanceof EnumType and
+    TaintTracking::localTaint(DataFlow::exprNode(fa), returnNode)
+  |
+    result = asTaintModel(api, "Argument[-1]", asOutput(api, returnNode))
+  )
+}
+
+string asOutput(Callable api, ReturnNodeExt node) {
+  if node.getKind() instanceof ValueReturnKind
+  then result = "ReturnValue"
+  else
+    result = parameterAccess(api.getParameter(node.getKind().(ParamUpdateReturnKind).getPosition()))
+}
+
+class FieldAssignment extends AssignExpr {
+  FieldAssignment() { exists(Field f | f.getAnAccess() = this.getDest()) }
+}
+
+class ParameterToFieldConfig extends TaintTracking::Configuration {
+  ParameterToFieldConfig() { this = "ParameterToFieldConfig" }
+
+  override predicate isSource(DataFlow::Node source) {
+    source instanceof DataFlow::ParameterNode and
+    isRelevantType(source.getType())
+  }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(FieldAssignment a |
+      a.getSource() = sink.asExpr() and
+      a.getDest().(VarAccess).getVariable().getCompilationUnit() =
+        sink.getEnclosingCallable().getCompilationUnit()
```

Comment:
Restricting the compilation unit doesn't seem right to me - that doesn't really guarantee that the field belongs to the right class. Use the Java type system. Also, might as well use `(FieldAccess).getField()` instead of `(VarAccess).getVariable()` for clarity here.

Reply:
Used `getDeclaringType` now instead (which is what this was intended to use). Let me know if you see other things we could restrict it to.